In [1]:
import sys, site
print("python:", sys.executable)
print("version:", sys.version)
print("site-packages:", site.getsitepackages() if hasattr(site, "getsitepackages") else "n/a")

python: C:\Users\user\Desktop\Model-Context-Protocol-102\.venv\Scripts\python.exe
version: 3.12.3 (tags/v3.12.3:f6650f9, Apr  9 2024, 14:05:25) [MSC v.1938 64 bit (AMD64)]
site-packages: ['C:\\Users\\user\\Desktop\\Model-Context-Protocol-102\\.venv', 'C:\\Users\\user\\Desktop\\Model-Context-Protocol-102\\.venv\\Lib\\site-packages']


In [2]:
import sys
sys.executable

'C:\\Users\\user\\Desktop\\Model-Context-Protocol-102\\.venv\\Scripts\\python.exe'

# 📘 Model Context Protocol 102 - Jupyter Notebook Tutorial

## 🛠️ Introduction
Welcome to **Model Context Protocol 102**! 🎯 This tutorial will guide you step-by-step in setting up and making API requests to MCP using Python inside a Jupyter Notebook.

By the end of this tutorial, you'll be able to:
- Understand how to interact with MCP APIs.
- Send GET requests and process responses.
- Handle errors and improve API interactions.

---

## 📌 Prerequisites
Before starting, make sure you have:
✅ Python 3.x installed  
✅ `venv` activated  
✅ Required dependencies installed (`pip install -r requirements.txt`)  
✅ Jupyter Notebook running (`jupyter notebook`)


---
## 📂 Notebook Structure
This tutorial follows a structured approach:

1️⃣ **Setup Environment** – Install required libraries and configure the API.  
2️⃣ **Make an API Request** – Perform a GET request to MCP and process the response.  
3️⃣ **Handle Errors & Debugging** – Learn best practices for handling API errors.  
4️⃣ **Next Steps** – Ideas for extending the tutorial.

## 1️⃣ Setup: Install & Import Dependencies
Before we start making API requests, let's ensure all required libraries are installed and imported.

```python
# Install required packages (if not installed)
!pip install requests
```

```python
# Import necessary libraries
import requests
import json
```




---

## 2️⃣ Making an API Request
Now, let's make a simple **GET request** to MCP. Replace `<YOUR_API_KEY>` with your actual API key.

```python
# Define API endpoint and headers
API_URL = "https://api.mcp.example.com/data"
HEADERS = {
    "Authorization": "Bearer <YOUR_API_KEY>",
    "Content-Type": "application/json"
}

# Make GET request
try:
    response = requests.get(API_URL, headers=HEADERS)
    response.raise_for_status()  # Raise error for bad status codes
    data = response.json()
    print("✅ Success! Data received:")
    print(json.dumps(data, indent=2))
except requests.exceptions.RequestException as e:
    print(f"❌ API Request failed: {e}")
```

---

## 3️⃣ Handling Errors & Debugging
It's important to handle errors gracefully. Let's add **error handling** and **logging**.

```python
def fetch_mcp_data():
    try:
        response = requests.get(API_URL, headers=HEADERS)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.HTTPError as http_err:
        print(f"🔴 HTTP error occurred: {http_err}")
    except requests.exceptions.ConnectionError:
        print("🔴 Connection error. Check your internet connection.")
    except requests.exceptions.Timeout:
        print("🔴 Request timed out. Try again later.")
    except Exception as err:
        print(f"🔴 Other error occurred: {err}")
    return None

# Run function
data = fetch_mcp_data()
if data:
    print("✅ Data successfully retrieved!")
```

---

## 4️⃣ Next Steps
Congratulations! 🎉 You've successfully:
- Made API requests to MCP.
- Processed JSON responses.
- Implemented error handling.

### 🔥 Explore More
✅ Modify the request to send **POST data**.  
✅ Implement **pagination** if API responses are large.  
✅ Use **environment variables** to store API keys securely.

---

## 📝 Summary
You've now built a **basic but functional** API request flow using Python and Jupyter Notebook. Keep experimenting and exploring!

Happy coding! 🚀
